pkg/types/chain_reader.go

```diff
@@ -51,21 +83,260 @@ type BoundContract struct {
+	return Expression{
+		Primitive: &AddressFilter{Addresses: addresses},
+	}
+}
+
+func (f *AddressFilter) Accept(visitor Visitor) {
+	visitor.VisitAddressFilter(*f)
+}
+
+type Confirmations int32
+
+const (
+	Finalized   = Confirmations(0)
+	Unconfirmed = Confirmations(1)
+)
+
+// ConfirmationsFilter is a primitive of QueryFilter that filters search to results that have a certain level of confirmation.
+// Confirmations map to different concepts on different blockchains.
+type ConfirmationsFilter struct {
+	Confirmations
+}
+
+func NewConfirmationsPrimitive(confs Confirmations) Expression {
+	return Expression{
+		Primitive: &ConfirmationsFilter{Confirmations: confs},
+	}
+}
+
+func (f *ConfirmationsFilter) Accept(visitor Visitor) {
+	visitor.VisitConfirmationFilter(*f)
+}
+
+// TimestampFilter is a primitive of QueryFilter that filters search in comparison to timestamp.
+type TimestampFilter struct {
+	Timestamp uint64
+	Operator  ComparisonOperator
+}
+
+func NewTimestampPrimitive(timestamp uint64, operator ComparisonOperator) Expression {
+	return Expression{
+		Primitive: &TimestampFilter{timestamp, operator},
+	}
+}
+
+func (f *TimestampFilter) Accept(visitor Visitor) {
+	visitor.VisitTimestampFilter(*f)
+}
+
+// TxHashFilter is a primitive of QueryFilter that filters search to results that contain txHash.
+type TxHashFilter struct {
+	TxHash string
+}
+
+func NewTxHashPrimitive(txHash string) Expression {
+	return Expression{
+		Primitive: &TxHashFilter{txHash},
+	}
+}
+
+func (f *TxHashFilter) Accept(visitor Visitor) {
+	visitor.VisitTxHashFilter(*f)
+}
+
+// Where eg. usage:
+// queryFilter, err := Where(
+//
+//		NewTxHashPrimitive("0xHash"),
+//		NewBooleanExpression("OR",
+//			NewBlockPrimitive(startBlock, Gte),
+//			NewBlockPrimitive(endBlock, Lte)),
+//		NewBooleanExpression("AND",
+//			NewBooleanExpression("OR",
+//				NewTimestampPrimitive(someTs1, Gte),
+//				NewTimestampPrimitive(otherTs1, Lte)),
+//			NewBooleanExpression("OR",(endBlock, Lte)),
+//				NewTimestampPrimitive(someTs2, Gte),
+//				NewTimestampPrimitive(otherTs2, Lte)))
+//	   )
+//	if err != nil{return nil, err}
+//
+// QueryKey(key, queryFilter)...
+func Where(expressions ...Expression) (QueryFilter, error) {
+	for _, expr := range expressions {
+		if !expr.IsPrimitive() {
+			if len(expr.BooleanExpression.Expressions) < 2 {
+				return QueryFilter{}, fmt.Errorf("all boolean expressions should have at least 2 expressions")
+			}
+		}
+	}
+	return QueryFilter{expressions}, nil
 }
 
-type EventFilter struct {
-	AddressList []string   // contract address
-	KeysList    [][]string // 2D list of indexed search keys, outer dim = AND, inner dim = OR.  Params[0] is the name of the event (or "event type"), rest are any narrowing parameters that may help further specify the event
-	Retention   time.Duration
+type Visitor interface {
+	VisitAddressFilter(filter AddressFilter)
+	VisitBlockFilter(filter BlockFilter)
```

Comment:
Maybe I'm confused about how this works, but it seems to me that we need a `KeyFilter` here somewhere, analogous to `KeysList`.

Also, what happens to `Retention`?  We need both of those for `RegisterEventFilter()` or whatever it's going to be named now (`RegisterSequenceFilter`?)

Reply:
I don't know what you mean by `KeysList`.  Keys are kept as a separate concept from QueryFilter. This is why there are multiple different QueryKey methods. I see key as a way to point to a result set which then gets filtered by QueryFilter. 
Lets give an evm example.
- QueryKey() would be used to say give me all sequences that belong to a "usdcTransferKey", which then is 
  narrowed down by `QueryFilter` block range, time range ...
- `QueryByKeyValuesIn` would be used to say give me all sequences where "usdcTransfer-value" is in [1000, 2000...], which returns all usdc transfer sequences where value is in xy, which then is narrowed down by `. 
  QueryFilter` block range, time range ...
- QueryByKeyValueEquality would be used to say give me all sequences where "usdcTransfer-value" is 
   [{500, >=, {1500, <=}] or just[{500, >=], which returns all usdc transfer sequences where value is in some range, 
    which then is narrowed down by `QueryFilter` block range, time range ...

Reply:
I can see `QueryByKeyValuesIn()` and `QueryByKeyValueEquality()` somehow being merged as they are doing a similar thing, but I haven't yet found a way to express it through proto in a clean way.

Reply:
I will add Retention, just haven't started on that yet. Retention would be tied to a key during binding/registering

Reply:
> I merged QueryByKeyValuesIn() and QueryByKeyValueEquality() into QueryByKeyValuesComparison()

Nice, thanks!

> I don't know what you mean by KeysList

I'm referring to the field of `EventFilter` we defined in the CLIP:
https://github.com/smartcontractkit/CLIPs/tree/2023-06-31-chain-reader/clips/2023-08-25-chain-reader
```
type EventFilter struct {
    AddressList []string    // contract address
    KeysList    [][]string  // 2D list of indexed search keys, outer dim = AND, inner dim = OR.  Params[0] is the name of the event (or "event type"), rest are any narrowing parameters that may help further specify the event
    Retention   time.Duration
}

type EventQuery struct{
    FromBlock      BlockID // either a hash or a big int
    ToBlock        BlockID // either a hash or a big int
    Filter         EventFilter
    Pending        bool
}

type Event struct {
    ChainId             string
	EventIndexInBlock   string
	Address             string
    TxHash              string
    BlockHash           string
    BlockNumber         int64
    BlockTimestamp      time.Time
    CreatedAt           time.Time
    Keys                []string
    Data                []byte
}
```
In the original design, `EventQuery` was what gets passed to `QueryEvents()` and a list of `Event`'s was what gets returned.  `EventQuery` filtered by address, keys, and block numbers... where Key[0] would usually be the "event type" or "event sig" and the rest of the keys would be topics or indices, or whatever else is needed to narrow things down.

It seems like you've taken that single method and split it in two, so that `Key[0]` can be treated differently from the rest of the keys. The disadvantage of that is having two separate methods, one for filtering just on `Key[0]` and one for filtering on any subset of keys including `Key[0]`.  I'm not sure I see what the advantage is of splitting it in two, is it motivated by trying to make the interface more user friendly, or is there another reason behind it?

Reply:
It is motivated by the ux. If you merge it into one function then all of the complexity lives in the QueryFilter, which means that there are a lot of invalid QueryFilters that the caller can come up with. I originally tried to make it into one function but the impl. is a lot more complex and it is a worse ux for the end user.

Reply:
If you merge `QueryKey()` and `QueryByKeyValuesComparison()` into a single method then the key value filtering would have to live in `QueryFilter`. This means that we would have to pass a slice of `QueryFilter` to `QueryKeys` since each `QueryFilter` would become bound to a key.

Reply:
I am trying to figure out the best UX, but I think that pushing this complexity into the filter could be more intuitive than this

Reply:
After revisiting this I think it may make more sense to merge these into one function. Even if the UX is worse it makes sense to allow more granular filtering which makes the issue of having to deal with a slice of `query.FIlter` a necessity.

Reply:
I might be missing something here, but I don't see why we should need to pass a slice of filters. We should only be passing a single filter which is the root node of a tree of filters constructed with the DSL.

Each `KeyFilter` primitive would either be for a single key, or for a collection of key/value pairs.  I could see pros and cons for either of those approaches.  But either way, the way to aggregate different filters is not through slices but through trees of logical operations. That's already how it works for all the other primitives right? Why should it be different for the `KeyFilter` primitive?

Reply:
I know things have evolved somewhat from here, but this section of that Rationale doc I wrote has some pseudo-code an examples of the general idea I have in mind:
https://smartcontract-it.atlassian.net/wiki/spaces/BCF/pages/583729325/Rationale+Behind+ChainReader+Design

I'm not sure how different that is from what you have now. If it's completely different then don't worry about it, but if there is a way to adjust what you have to go more in this direction I think it might make for a cleaner interface.